src/main/java/org/openpnp/spi/base/AbstractCamera.java

```diff
@@ -253,16 +296,172 @@ public Location getLocation(HeadMountable tool) {
         return super.getLocation();
     }
 
+    @Override
+    public Location getActualLocation() {
+        Location location = getLocation();
+        LengthUnit units = location.getUnits();
+        //Check for any virtual axis and replace them with their home coordinate
+        Axis axis = getAxisX();
+        if ((axis != null) && (axis.getClass() == ReferenceVirtualAxis.class)) {
```

Comment:
Checks such as these

` if ((axis != null) && (axis.getClass() == ReferenceVirtualAxis.class)) {`

should be

` if (axis instanceof ReferenceVirtualAxis) {`

Note, `instanceof` includes the `null` test.

Reply:
**Get Actual Location**

1. Yes, ActualLocation() does not need to know anything about the true image plane (that is where the image is in perfect focus) but it does need to be fixed in relation to the physical location of the camera.
2. Through 5. Yes, that all makes sense to me, I'll see if I can make those changes.

**Camera interface**
Ok, I can do that.

**Use Cases**
Yes, eventually all the places where getUnitsPerPixel() is called should be reviewed to see if a known height can be passed (that part of the future upgrades I mentioned).  Before then, they should continue to work as they have in the past.

As far as the Auto Camera View I'll let you decide what is best there since I haven't tried it yet.

**Branch to merge**
Yes, I soon as I make the suggested updates I think it could be merged into **test**

Reply:
@markmaker,

I'm running into problems moving the new methods from the Camera interface to AbstractCamera.  There seems to be several places where Camera is used where I need access to at least some of those methods.  Should I change as many as I need to AbstractCamera (which may lead to a big chain of changes) or should I leave them as I originally had?  Also you didn't mention getUnitsPerPixel(Length z) and getUnitsPerPixel(Location location) should those be moved as well?

Oh, and I forgot to previously mention - yes to instanceof.

Tony

Reply:
As far as I've understood, `getUnitsPerPixel(Length z)` and `getUnitsPerPixel(Location location)` are the real interface. They are valid additions to `Camera`. The others should be internal. 

I've seen the use in `isReticleHeightChangable()` but this can equally be expressed in comparing `getUnitsPerPixel(Length z)` at two arbitrary `z`, right?

Other uses?

_Mark

Reply:
If I use your suggestion to move things out of Camera into AbstractCamera then don't I need to change CameraView.setCamera(Camera camera) to CameraView.setCamera(AbstractCamera camera) because at least some of the new methods are used in CameraView?  And then I think there are several of places in CameraPanel which will need to change from Camera to AbstractCamera.  Is that what you really want or am I missing something?

Tony

Reply:
I guess `getDefaultZ()` must be added to `Camera` too. Other than that, what are you missing?

_Mark